stm32wbax 802154 driver support

vtardy-st opened the pull request:
Add support of 802.15.4 driver for STM32WBA
Update Kconfig/CMakeLists/dts/bindings related to 802.15.4 support for STM32WBA
In soc/stm32, update files to support of 802.15.4 on STM32WBA
Enable ieee802154 node for nucleo_wba65ri board

vtardy-st committed:
drivers: ieee802154: add support of 802.15.4 for STM32WBA

Add the driver itself and Kconfig/CMakeLists/dts/bindings related to it.
Other files and libraries needed are in ST's dedicated folder
hal_stm32 (modules/hal/stm32).

Signed-off-by: Vincent Tardy <vincent.tardy@st.com>

vtardy-st committed:
soc: stm32: add 802.15.4 support for STM32WBA

Add 802.15.4 support in hci_if files

Signed-off-by: Vincent Tardy <vincent.tardy@st.com>

msmttchr reviewed (changes requested):
@vtardy-st, I have submitted a couple of comments, but I have realized that possibly a rebase is needed, before further review. Could you please rebase the code and re-submit?

HoZHel commented on boards/st/nucleo_wba65ri/nucleo_wba65ri.dts:
```diff
@@ -108,13 +108,13 @@
 &rtc {
 	status = "okay";
-	clocks = <&rcc STM32_CLOCK(APB7, 21)>,
+	clocks = <&rcc STM32_CLOCK_BUS_APB7 0x00200000>,
```
I think using the macro is a better choice. Is there any reason you are changing it?

HoZHel commented on boards/st/nucleo_wba65ri/nucleo_wba65ri.dts:
```diff
@@ -108,13 +108,13 @@
 
 &usart1 {
-	clocks = <&rcc STM32_CLOCK(APB2, 14)>,
+	clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00004000>,
```
The same here.

HoZHel commented on boards/st/nucleo_wba65ri/nucleo_wba65ri.dts:
```diff
@@ -150,7 +140,7 @@
 
 stm32_lp_tick_source: &lptim1 {
-	clocks = <&rcc STM32_CLOCK(APB7, 11)>,
+	clocks = <&rcc STM32_CLOCK_BUS_APB7 0x00000800>,
```
Also here.

HoZHel commented on soc/st/stm32/stm32wbax/hci_if/IEEE802154/802154_plat.c:
```diff
@@ -0,0 +1,102 @@
+void assert_failed(uint8_t* file, uint32_t line)
+{
+  /* Prevent unused argument(s) compilation warning */
+  (void)file;
+  (void)line;
```
```suggestion
  ARG_UNUSED(file);
  ARG_UNUSED(line);
```

msmttchr reviewed (commented):
Could you please explore the possibility of "merging" hci_if/ble and hci_if/IEEE802154 directories ? I think they perform similar functions.

vtardy-st commented on boards/st/nucleo_wba65ri/nucleo_wba65ri.dts:
```diff
@@ -165,6 +155,22 @@ stm32_lp_tick_source: &lptim1 {
+		scratch_partition: partition@1b0000 {
+			label = "image-scratch";
+			reg = <0x001b0000 DT_SIZE_K(64)>;
+		};
```
Code is removed

mathieuchopstm commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1079 @@
+{
+	LOG_DBG("Fetching capabilities at boot");
+
+	stm32wba_data.capabilities =
+		IEEE802154_HW_ENERGY_SCAN |
+		IEEE802154_HW_FCS |
+		IEEE802154_HW_FILTER |
+		IEEE802154_HW_PROMISC |
+#if CONFIG_IEEE802154_STM32WBA_CSMA_CA_ENABLED
+		IEEE802154_HW_CSMA |
+#endif
+		IEEE802154_HW_TX_RX_ACK |
+		IEEE802154_HW_RETRANSMISSION |
+		IEEE802154_HW_RX_TX_ACK |
+#if (CONFIG_MAC_CSL_TRANSMITTER_ENABLE == 1)
+		IEEE802154_HW_TXTIME |
+#endif
+#if (CONFIG_MAC_CSL_RECEIVER_ENABLE == 1)
+		IEEE802154_HW_RXTIME |
+#endif
+		IEEE802154_HW_SLEEP_TO_TX |
+		IEEE802154_RX_ON_WHEN_IDLE
+		;
```
`stm32wba_data` should just be initialized with this already set, and this function removed.

mathieuchopstm commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1079 @@
+
+/* Radio device API */
+
+static enum ieee802154_hw_caps stm32wba_get_capabilities(const struct device *dev)
+{
+	return stm32wba_data.capabilities;
+}
```
```suggestion
static enum ieee802154_hw_caps stm32wba_get_capabilities(const struct device *dev)
{
	struct stm32wba_802154_data *data = STM32WBA_802154_DATA(dev);

	return data->capabilities;
}
```

(or remain single instance, but in this case the whole `stm32wba_data.capabilities` should be removed - just return the value computed by `stm32wba_get_capabilities_at_boot` here, since that's the only place where it is used(?))

vtardy-st replied:
stm32wba_get_capabilities() directly returns mask value previously present in stm32wba_get_capabilities_at_boot().
capabilities field is removed in the stm32wba_802154_data structure

mathieuchopstm reviewed (commented):
first pass

etienne-lms commented on soc/st/stm32/stm32wbax/CMakeLists.txt:
```diff
@@ -10,29 +10,13 @@ zephyr_sources_ifdef(CONFIG_PM
+zephyr_sources(hci_if/sys_wireless_plat.c)
+zephyr_sources(hci_if/host_stack_if.c)
+zephyr_sources(hci_if/ll_sys_if_adapt.c)
+zephyr_sources(hci_if/Utilities/tim_serv/stm32_timer.c)
```
could you indent each line by 2 space char, for consistency along CMakeLists.txt files.

vtardy-st replied:
suggestion implemented

etienne-lms reviewed (commented):
Some suggestions, in hope it helps. Discard them if they are not relevant.

Commit *"drivers: ieee802154: added support of 802.15.4 for STM32WBA (part 2)"* seems a fixup commit for commit *"drivers: ieee802154: added support of 802.15.4 for STM32WBA"*. Can't the 2 commits be squashed? (I mean have the former commit sufficient).

I don't understand why you need a preparatory commit, adding stuff, then a later commit  ("Merging of ...") factorizing code and removing almost all content from the preparatory commit. For simplicity, couldn't you have a commit series like:
1. "drivers: ieee802154: added support of 802.15.4 for STM32WBA"
2. "soc: stm32: added timer service for 802.15.4 support for STM32WBA"
3. "soc: stm32: added soc files for 802.15.4 support for STM32WBA"

By the way, please replace past tense (like "added support ...") with imperative mode ("add support ...") in the commit messages. Past tense usually denotes things implemented before your change/commit.

vtardy-st commented on drivers/ieee802154/Kconfig.stm32wba:
```diff
@@ -0,0 +1,77 @@
+	  Indicate if CSMA-CA feature is supported.
+
+config IEEE802154_STM32WBA_RX_STACK_SIZE
+	int "Driver's internal RX thread stack size"
```
I have implemented your suggestion

vtardy-st commented on drivers/ieee802154/Kconfig.stm32wba:
```diff
@@ -0,0 +1,77 @@
+	  a lot of log records in a stress environment.
+
+config NET_L2_CUSTOM_IEEE802154_STM32WBA
+	bool "ST MAC"
```
I have implemented your suggestion

vtardy-st commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1079 @@
+static void stm32wba_irq_config(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+}
```
This function has been removed

vtardy-st commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1079 @@
+	void (*irq_config_func)(const struct device *dev);
+};
+
+static struct stm32wba_802154_data stm32wba_data;
```
STM32WBA_802154_DATA() has been removed, stm32wba_data is directly called

vtardy-st commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1079 @@
+#define NSEC_PER_TEN_SYMBOLS (10 * IEEE802154_PHY_OQPSK_780_TO_2450MHZ_SYMBOL_PERIOD_NS)
+
+/* Convenience defines for RADIO */
+#define STM32WBA_802154_DATA(dev) ((struct stm32wba_802154_data * const)(dev)->data)
```
STM32WBA_802154_DATA() has been removed

vtardy-st commented on drivers/ieee802154/Kconfig.stm32wba:
```diff
@@ -0,0 +1,70 @@
+	  Number of buffers in 802.15.4 driver serialization host's receive queue.
+	  Increase this number if you are expected to be in a noisy environment, handling many
+	  end-devices and you have sufficient RAM to allocate memory to buffers.
+	  If this value is increased, IEEE802154_STM32WBA_RX_STACK_SIZE value might required 
```
suggestion implemented

vtardy-st commented on drivers/ieee802154/Kconfig.stm32wba:
```diff
@@ -0,0 +1,70 @@
+	help
+	  This option sets the driver's stack size for its internal RX thread.
+	  The default value should be sufficient, but in case it proves to be
+	  a too little one, this option makes it easy to play with the size.
```
suggestion implemented

vtardy-st commented on drivers/ieee802154/Kconfig.stm32wba:
```diff
@@ -0,0 +1,70 @@
+	bool "Frame reception failures logging"
+	help
+	  There are few cases where the frame reception failure can happen because of
+	  internal cause. These cases are reported forward by general code error.
```
suggestion implemented

vtardy-st commented on soc/st/stm32/stm32wbax/hci_if/sys_wireless_plat.c:
```diff
@@ -29,7 +30,7 @@ struct entropy_stm32_rng_dev_data {
 	struct stm32_pclken *pclken;
 };
-
+#if defined(CONFIG_BT_STM32WBA)
```
suggestion implemented

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1051 @@
+	switch (type) {
+	case IEEE802154_STM32WBA_CONFIG_CCA_THRESHOLD:
+		LOG_DBG("Setting CCA_THRESHOLD: %d", config->cca_thr);
+		if (st_ral_set_cca_energy_detect_threshold(config->cca_thr)) {
```
If `st_ral_set_cca_energy_detect_threshold()` does not return an *essentially bolean* value, please compare against 0 (or any relevant value).

Ditto for the other calls to `st_ral_...()` with a return value.

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1051 @@
+
+	st_ral_set_max_csma_backoff(MAX_CSMA_BACKOFF);
+	st_ral_set_max_frame_retries(MAX_FRAME_RETRY);
+	st_ral_set_cca_energy_detect_threshold(CCA_THRESHOLD);
```
Return code is expected to be tested.

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1051 @@
+}
+
+static int stm32wba_attr_get_extended(enum ieee802154_stm32wba_attr attr,
+				      const struct ieee802154_stm32wba_attr_value *value)
```
It seems strange a `..._get...()` function with such an output argument declares the argument pointer as pointer to const. I believe it should not and the several casts below should be removed.

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1051 @@
+		break;
+
+	default:
+		LOG_ERR("Unsupported attribute: %u", (uint32_t)attr);
```
Cast not needed.

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1051 @@
+		IEEE802154_HW_TX_RX_ACK |
+		IEEE802154_HW_RETRANSMISSION |
+		IEEE802154_HW_RX_TX_ACK |
+#if (CONFIG_MAC_CSL_TRANSMITTER_ENABLE == 1)
```
I can't see this configuration option defined anywhere. Neither `CONFIG_MAC_CSL_RECEIVER_ENABLE` used 2 lines below.

Note that for config options, the related `CONFIG_xxx` macro is defined when the option is enabled and not defined when the option is disabled. Therefore here you should use `#ifdef`.

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1051 @@
+{
+	ARG_UNUSED(dev);
+
+	if (channel < drv_attr.phy_channel_range.from_channel
+	 || channel > drv_attr.phy_channel_range.to_channel) {
```
```suggestion
	if (channel < drv_attr.phy_channel_range.from_channel ||
	    channel > drv_attr.phy_channel_range.to_channel) {

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1051 @@
+		return -ENOTSUP;
+	}
+
+	if (type == IEEE802154_FILTER_TYPE_IEEE_ADDR) {
```
switch/case?

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1051 @@
+
+	ack_pkt = net_pkt_rx_alloc_with_buffer(stm32wba_radio->iface, ack_len,
+					       AF_UNSPEC, 0, K_NO_WAIT);
+	if (!ack_pkt) {
```
```suggestion
	if (ack_pkt == NULL) {

vtardy-st committed:
boards: st: nucleo_wba65ri: enable ieee802154 node

enable ieee802154 node for nucleo_wba65ri

Signed-off-by: Vincent Tardy <vincent.tardy@st.com>

msmttchr reviewed (approved):
LGTM from what I can understand. Thanks

erwango reviewed (changes requested):
Please update https://github.com/zephyrproject-rtos/zephyr/blob/6835e0ed3107012026de4eb0d7fa3ba1ebed2820/MAINTAINERS.yml#L4370 and https://github.com/zephyrproject-rtos/zephyr/blob/6835e0ed3107012026de4eb0d7fa3ba1ebed2820/MAINTAINERS.yml#L4352

AhmedFaizyST commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1035 @@
+
+	link_layer_register_isr();
+#if defined(CONFIG_NET_L2_CUSTOM_IEEE802154_STM32WBA)
+	ll_sys_mac_cntrl_init();
```
This call to ll_sys_mac_cntrl_init() should be moved to be after calling ieee802154_init() because the MAC initialization needs the mac_port to be initialized first.

vtardy-st replied:
change has been implemented

AhmedFaizyST commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1035 @@
+
+#define LOG_MODULE_NAME ieee802154_stm32wba
+
+#define LOG_LEVEL CONFIG_IEEE802154_DRIVER_LOG_LEVEL
```
This macro definition will cause a compilation error when CONFIG_LOG=n. It should be defined as LOG_LEVEL_NONE in case CONFIG_IEEE802154_DRIVER_LOG_LEVEL is not defined.

#if defined(CONFIG_IEEE802154_DRIVER_LOG_LEVEL)
#define LOG_LEVEL CONFIG_IEEE802154_DRIVER_LOG_LEVEL
#else
#define LOG_LEVEL LOG_LEVEL_NONE
#endif

etienne-lms replied:
I've fixed my proposal above: idea is to use `_LOG_LEVEL_RESOLVE()`.
Actually, I think `LOG_MODULE_REGISTER()` could process its 2nd arg with `_LOG_LEVEL_RESOLVE()` in a generic way, but that's out of the scope of the current P-R.

vtardy-st commented on drivers/ieee802154/Kconfig.stm32wba:
```diff
@@ -0,0 +1,69 @@
+	help
+	  This option enables the CSMA-CA feature.
+
+config IEEE802154_STM32WBA_RX_STACK_SIZE
+	int "Driver RX thread stack size"
+	default 800
+	help
+	  This option sets the driver's stack size for its internal RX thread.
+	  The default value should be sufficient, but in case it proves to be
+	  too low, adjust as needed.
```
change implemented

vtardy-st commented on soc/st/stm32/stm32wbax/Kconfig.defconfig:
```diff
@@ -50,6 +50,31 @@ config SYSTEM_WORKQUEUE_STACK_SIZE
 
+config IEEE802154_STM32WBA
+	select USE_STM32_HAL_CORTEX
+	select USE_STM32_HAL_RAMCFG
```
USE_STM32_HAL_CORTEX and USE_STM32_HAL_RAMCFG have been removed

vtardy-st commented on drivers/ieee802154/Kconfig.stm32wba:
```diff
@@ -0,0 +1,69 @@
+	help
+	  Custom STM32WBA series IEEE 802.15.4 L2 Implementation.
+
+endif
```
implemented

vtardy-st commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1041 @@
+		uint16_t short_addr = (uint16_t)config->ack_fpb.addr[0] |
+				     (uint16_t)(config->ack_fpb.addr[1] << 8);
+
+		if (st_ral_pending_bit_for_short_addr_set((const uint16_t)short_addr)) {
```
cast removed

vtardy-st commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1041 @@
+		uint16_t shortAddr = (uint16_t)config->ack_fpb.addr[0] |
+				     (uint16_t)(config->ack_fpb.addr[1] << 8);
+
+		if (st_ral_pending_bit_for_short_addr_clear((const uint16_t)shortAddr)) {
```
cast removed

vtardy-st commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1041 @@
+				     const struct ieee802154_config *config)
+{
+	ARG_UNUSED(dev);
+	int ret = 0;
```
implemented

vtardy-st commented on drivers/ieee802154/ieee802154_stm32wba.h:
```diff
@@ -0,0 +1,254 @@
+#ifndef ZEPHYR_DRIVERS_IEEE802154_IEEE802154_STM32WBA_H_
+#define ZEPHYR_DRIVERS_IEEE802154_IEEE802154_STM32WBA_H_
+
+#include <zephyr/net/ieee802154_radio.h>
```
#include <zephyr/kernel.h> added

asm5878 commented on soc/st/stm32/stm32wbax/hci_if/sys_wireless_plat.c:
I totally agree to rename this file.

We have a subset of these functions to be reimplemented  according Zephyr platform.

But we have the rest of functions implemented though bare metal code.
We should mark in a integration doc/readme where these functions are coming from to facilitate the maintenance.
"bleplat.c" is the name used in ST proprietary BLE applications.

asm5878 reviewed (dismissed):
More formal improvements could be done in a  next PR.
For what I can see, good enough to be approved.

mathieuchopstm reviewed (dismissed):
Overall acceptable as said by @asm5878.

etienne-lms reviewed (commented):
The overall LGTM but I'll have a second look...

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1075 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32wba_ieee802154
+
+#define LOG_MODULE_NAME ieee802154_stm32wba
```
These macros are not needed before the header files inclusion.
Could you move them after?

vtardy-st commented on drivers/ieee802154/CMakeLists.txt:
```diff
@@ -20,4 +20,5 @@ zephyr_library_sources_ifdef(CONFIG_IEEE802154_MCXW		ieee802154_mcxw.c		ieee8021
 zephyr_library_sources_ifdef(CONFIG_IEEE802154_NRF5		ieee802154_nrf5.c)
 zephyr_library_sources_ifdef(CONFIG_IEEE802154_RF2XX		ieee802154_rf2xx.c)
 zephyr_library_sources_ifdef(CONFIG_IEEE802154_RF2XX		ieee802154_rf2xx_iface.c)
+zephyr_library_sources_ifdef(CONFIG_IEEE802154_STM32WBA	ieee802154_stm32wba.c)
```
tab added

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1071 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_LEVEL _LOG_LEVEL_RESOLVE(CONFIG_IEEE802154_DRIVER_LOG_LEVEL)
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL);
```
Could you move `#define LOG_LEVEL ...` and `LOG_MODULE_REGISTER(...)` down after the header files inclusion.

vtardy-st replied:
suggestion implemented

etienne-lms commented on drivers/ieee802154/ieee802154_stm32wba.c:
```diff
@@ -0,0 +1,1071 @@
+#include <zephyr/debug/stack.h>
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/net_pkt.h>
+#include <zephyr/sys/byteorder.h>
```
Ntpicking biut it's easier to maintain when ordered alphabetically (or at least ordered with some rational criteria), e.g.:
```c
#include <zephyr/arch/cpu.h>
#include <zephyr/debug/stack.h>
#include <zephyr/device.h>
#include <zephyr/init.h>
#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>
#include <zephyr/net/net_if.h>
#include <zephyr/net/net_pkt.h>
#include <zephyr/sys/byteorder.h>

etienne-lms reviewed (commented):
LGTM with 2 minor comments.

etienne-lms reviewed (dismissed):
LGTM

erwango reviewed (changes requested):
Please also update PR description with some details on what is added here.

erwango reviewed (changes requested):
Is there a way to build this in CI or isn't it possible due to the required blob ?

In any case, I think we should add a documentation bit in the board doc to explain that this is supported but require to fetch the blob first.

vtardy-st committed:
MAINTAINERS: Add driver ieee 802.15.4 for STM32 Wireless Platform

Add ieee802154 driver in STM32 Wireless Platform section

Signed-off-by: Vincent Tardy <vincent.tardy@st.com>

vtardy-st committed:
west.yml: update hal_stm32 version

Update hal_stm32 version to latest commit:
dc7c2543a079c57a4e7dbee54ed7877880dc235b

Signed-off-by: Vincent Tardy <vincent.tardy@st.com>

etienne-lms commented:
I'm not sure it's a good idea to have the manifest updated from its specific commit. At least the manifest update commit being after the commit enabling 802.15.4 on nucleo_wba65ri would also make that board to fail to build on that commit. Will earlier changes in the driver would make wba65 boards with BLE enabled to fail to build. 

I wonder if at least the commit that updates the manifest could be placed first in the series.

erwango commented:
@cfriedt, @jhedberg Last remaining "issue" of this PR are the points linked to MAINTAINERS update (not even sure of that one) and Binary Blob addition (which is not dealt by TSC anymore so I guess it's up to you).

cfriedt commented on drivers/ieee802154/Kconfig.stm32wba:
```diff
@@ -0,0 +1,77 @@
+	  the overall packet length while others not. Allow to configure this
+	  behavior, based on the selected upper layer.
+
+config IEEE802154_STM32WBA_DELAY_TRX_ACC
+	int "Clock accuracy for delayed operations"
+	default 20
+	range 0 255
+	help
+	  Accuracy of the clock used for scheduling radio delayed operations (delayed transmission
+	  or delayed reception), in ppm.
```
A possible better alternative is to use rate-limited logging and to log a count of how many rx errors there are.

cfriedt commented:
@erwango - looks generally fine to me. @nashif and @MaureenHelm can comment on the MAINTAINERS file.
